File: scripts/generate_panel_cards.py
"""Frame captured page panels onto branded 1080x1350 export cards.

THE CHART IS NEVER REDRAWN HERE. The court and the zone table are React
components rendered from the deployed payloads, and reimplementing either in
Python would be a second implementation that drifts from the site the first
time a bin edge or a zone label moves (the ADR-0009 rule, applied to
pixels). So the browser renders them — scripts/generate-panel-cards.ts
drives real Chromium over the real page — and this step only COMPOSES: a
captured PNG onto the card family's ground, under the wordmark, over the
hero's kicker.

Captures arrive with transparent backgrounds (Playwright omitBackground), so
the panel sits on the shared radial ground rather than on a visible rectangle
of a slightly different dark.

Driven by scripts/generate-panel-cards.ts; specs arrive as JSON on stdin:
  [{"slug", "capturePath", "kicker", "label", "outName"}]
Output: social-exports/panels/<slug>-<panel>.png — gitignored, like every
vertical export card.
"""

from __future__ import annotations

import sys
from pathlib import Path

from PIL import Image, ImageDraw

from card_common import (
    DISPLAY_900,
    SANS_500,
    WORDMARK,
    draw_tracked,
    load_font,
    radial_ground,
    read_specs,
    tracked_width,
)

CARD_W, CARD_H = 1080, 1350
MARGIN = 72
OUT_DIR = Path("social-exports/panels")

SITE = "nbagoodshots.com"


def compose(spec: dict, sans_woff: Path) -> Image.Image:
    # An aspect-matched sampling grid: the share card's historical grid is
    # tuned to a landscape frame and would smear across a portrait one.
    card = radial_ground((CARD_W, CARD_H), small=(240, 300)).convert("RGBA")

    wordmark = Image.open(WORDMARK).convert("RGBA")
    wm_h = 40
    wm = wordmark.resize(
        (round(wordmark.width * wm_h / wordmark.height), wm_h),
        Image.Resampling.LANCZOS,
    )
    card.alpha_composite(wm, (MARGIN, MARGIN - 4))

    draw = ImageDraw.Draw(card)
    label_font = load_font(sans_woff, 18)
    draw_tracked(card, draw, spec["label"], label_font, MARGIN, 136, 18 * 0.22, 0.70)

    # Caption block, bottom-aligned: the hero's kicker over the site line.
    kicker_size = 20
    kicker_font = load_font(sans_woff, kicker_size)
    column = CARD_W - MARGIN * 2
    while (
        kicker_size > 12
        and tracked_width(spec["kicker"], kicker_font, kicker_size * 0.18) > column
    ):
        kicker_size -= 1
        kicker_font = load_font(sans_woff, kicker_size)
    site_font = load_font(sans_woff, 17)
    kicker_h, gap, site_h = 26, 12, 22
    caption_top = CARD_H - MARGIN - (kicker_h + gap + site_h)

    # The panel fills what the chrome leaves, contained (never cropped — a
    # cropped axis label is a lie the frame tells).
    capture = Image.open(spec["capturePath"]).convert("RGBA")
    region_top = 200
    region_h = caption_top - 48 - region_top
    scale = min(column / capture.width, region_h / capture.height)
    size = (round(capture.width * scale), round(capture.height * scale))
    panel = capture.resize(size, Image.Resampling.LANCZOS)
    card.alpha_composite(
        panel,
        (round((CARD_W - size[0]) / 2), region_top + round((region_h - size[1]) / 2)),
    )

    draw_tracked(
        card, draw, spec["kicker"], kicker_font, MARGIN, caption_top,
        kicker_size * 0.18, 0.85,
    )
    draw_tracked(
        card, draw, SITE, site_font, MARGIN, caption_top + kicker_h + gap,
        17 * 0.22, 0.55,
    )
    return card


def main() -> None:
    specs = read_specs(sys.stdin)
    for asset in (DISPLAY_900, SANS_500, WORDMARK):
        if not asset.exists():
            sys.exit(f"missing asset: {asset} (run npm install?)")

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    for spec in specs:
        capture = Path(spec["capturePath"])
        if not capture.exists():
            sys.exit(f"missing capture for {spec['slug']}: {capture}")
        out = OUT_DIR / spec["outName"]
        compose(spec, SANS_500).convert("RGB").save(out, format="PNG", optimize=True)
        print(f"panel {CARD_W}x{CARD_H} -> {out} ({out.stat().st_size:,} bytes)")


if __name__ == "__main__":
    main()

File: scripts/generate-panel-cards.ts
// Capture each hero page's court and zone table from the REAL page, then
// frame them as 1080x1350 export cards (ADR-0009's rule applied to pixels:
// the chart has one implementation, and it is the React component — so the
// export drives real Chromium over the real deployment rather than redrawing
// the court in Python).
//
// Usage:
//   npm run cards:panels
//   npm run cards:panels -- --slug shai-gilgeous-alexander
//
// Playwright is a devDependency and never ships: this touches the build
// output the same way a human with DevTools would. Output lands in
// social-exports/panels/ (gitignored) beside the vertical cards.

import { spawn, spawnSync } from 'node:child_process'
import { mkdirSync, rmSync } from 'node:fs'
import { resolve } from 'node:path'
import { chromium } from 'playwright'
import { SCAFFOLD_SENTINEL } from '../src/heroes/authoring'
import { HEROES } from '../src/heroes/registry'
import { canonicalSeasonOf } from '../src/heroes/types'

const PORT = 4179
const BASE = `http://localhost:${PORT}`
const CAPTURE_DIR = resolve('social-exports/_captures')

// Each panel names the selector it captures and the structural label the
// frame prints. Labels are direction-free section copy, never a per-hero
// claim, so they carry no verdict-guard obligations.
// Selectors are scoped to `.zone-section` deliberately: `.table-panel` and
// `.chart-panel` are the SHARED panel classes every act reuses (zone,
// creation, assisted makes, free throw), so an unscoped locator matches four
// elements and Playwright's strict mode rejects it. Scoping also means a
// future act cannot silently change which panel gets exported.
const PANELS = [
  {
    key: 'court',
    selector: '.zone-section .chart-panel',
    label: 'Shot making vs league average',
  },
  {
    key: 'table',
    selector: '.zone-section .table-panel',
    label: 'Zone by zone',
  },
]

const argv = process.argv.slice(2)
const only = argv.includes('--slug') ? argv[argv.indexOf('--slug') + 1] : undefined
const heroes = only === undefined ? HEROES : HEROES.filter((h) => h.slug === only)
if (heroes.length === 0) {
  console.error(`no registered hero with slug '${only}'`)
  process.exit(1)
}
const targets = heroes.filter((h) => !h.thesis.includes(SCAFFOLD_SENTINEL))
if (targets.length === 0) {
  console.log('no heroes ready for cards')
  process.exit(0)
}

// The preview server serves dist/, so build first — it is under a second and
// it guarantees the capture matches what actually deploys.
console.log('building...')
const built = spawnSync('npm', ['run', 'build'], { stdio: 'ignore', shell: true })
if (built.status !== 0) {
  console.error('build failed — run `npm run build` to see why')
  process.exit(1)
}

const server = spawn(
  'npx',
  ['vite', 'preview', '--port', String(PORT), '--strictPort'],
  { stdio: 'ignore', shell: true },
)
const shutdown = () => {
  if (server.killed || server.pid === undefined) return
  // On Windows `shell: true` means server.pid is the cmd.exe wrapper, and
  // killing it ORPHANS the vite process — which keeps the port bound and
  // holds stdout open, so the whole command hangs with no error visible.
  // Kill the tree.
  if (process.platform === 'win32') {
    spawnSync('taskkill', ['/pid', String(server.pid), '/T', '/F'], {
      stdio: 'ignore',
    })
  } else {
    server.kill()
  }
}
process.on('exit', shutdown)
process.on('SIGINT', () => {
  shutdown()
  process.exit(130)
})

const waitForServer = async () => {
  for (let i = 0; i < 60; i++) {
    try {
      const res = await fetch(BASE)
      if (res.ok) return
    } catch {
      // not up yet
    }
    await new Promise((r) => setTimeout(r, 500))
  }
  throw new Error(`preview server never came up on ${BASE}`)
}

const run = async () => {
  await waitForServer()
  rmSync(CAPTURE_DIR, { recursive: true, force: true })
  mkdirSync(CAPTURE_DIR, { recursive: true })

  const browser = await chromium.launch()
  // Wide enough that the zone table lays out at its desktop grain instead of
  // inside its horizontal scroller; 3x so the type survives the downscale.
  const context = await browser.newContext({
    viewport: { width: 1440, height: 1400 },
    // 2x, not 3x: the panels are placed ~936px wide, so 2x already oversamples,
    // and a 3x surface over a page this tall makes the table capture hang.
    deviceScaleFactor: 2,
    colorScheme: 'dark',
  })

  const specs: Record<string, string>[] = []
  for (const hero of targets) {
    const page = await context.newPage()
    await page.goto(`${BASE}/${hero.slug}`, { waitUntil: 'networkidle' })
    // The panels only exist once all four payloads have parsed, so this
    // doubles as the load gate. A hero whose data fails to load throws here
    // rather than silently exporting an empty frame.
    await page.waitForSelector('.chart-panel svg', { timeout: 30_000 })
    await page.waitForSelector('.table-panel', { timeout: 30_000 })
    // String form deliberately: this expression runs in the PAGE, and the
    // repo's tsc has no DOM lib for scripts/, so a callback referencing
    // `document` fails the build (TS2584).
    await page.evaluate('document.fonts.ready')
    // Interaction affordances are lies in a still: "Click any zone" is an
    // instruction a PNG cannot honor. Injected here rather than fixed in the
    // app, because the affordance is correct on the real page.
    await page.addStyleTag({ content: '.chart-hint-slot { visibility: hidden; }' })

    for (const panel of PANELS) {
      const capturePath = resolve(CAPTURE_DIR, `${hero.slug}-${panel.key}.png`)
      await page.locator(panel.selector).screenshot({
        path: capturePath,
        omitBackground: true,
        // Without these the capture waits on element stability forever
        // rather than failing, which is a hang with no message at 3am.
        animations: 'disabled',
        timeout: 20_000,
      })
      specs.push({
        slug: hero.slug,
        capturePath,
        kicker: canonicalSeasonOf(hero).kicker,
        label: panel.label,
        outName: `${hero.slug}-${panel.key}.png`,
      })
      console.log(`captured ${hero.slug} ${panel.key}`)
    }
    await page.close()
  }
  await browser.close()
  shutdown()

  const framed = spawnSync('python', ['scripts/generate_panel_cards.py'], {
    input: JSON.stringify(specs),
    stdio: ['pipe', 'inherit', 'inherit'],
  })
  if (framed.error) {
    console.error(`failed to run python: ${framed.error.message}`)
    process.exit(1)
  }
  process.exit(framed.status ?? 1)
}

run().catch((err) => {
  console.error(err instanceof Error ? err.message : String(err))
  shutdown()
  process.exit(1)
})
